[lldb][telemetry] Implement LLDB Telemetry (part 1)

oontvoo opened the pull request:
Details:
- This is a subset of PR/98528.( Pavel's suggestion was to split up the patch to make reviewing easier)
- This contains only the concrete implementation of the framework to be used but no usages yet.
- I plan to send a few follow-up patches:
  + part2 : includes changes in the plugin-manager to set up the plugin stuff (ie., how to create a default vs vendor impl)
  + part3 (all of the following can be done in parallel):
        * part 3_a: define DebuggerTelemetryInfo and related methods to collect data about debugger startup/exit
        * part 3_b: define TargetTelemetryInfo and related methods to collect data about debug target(s)
        * part 3_c: define CommandTelemetryInfo and related methods to collect data about debug-commands
        * part 3_d: define ClientTelemtryInfo and related methods to collect data about lldb-dap/any other client

Note: Please ignore all changes under llvm/.... These will be reverted after the pending LLVM patch is submitted.

labath commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,309 @@
+  KindType getKind() const override { return LldbEntryKind::BaseInfo; }
+
+  static bool classof(const TelemetryInfo *t) {
+    if (t == nullptr)
+      return false;
```
This isn't necessary. Null pointers are handled in the `or_null` cast variants.

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,338 @@
+  if (!Point.has_value())
+    return 0;
+
+  return Point.value().time_since_epoch().count();
```
This doesn't actually guarantee a result in nanoseconds (because the resolution of the steady_clock is system dependent). You want to explicitly convert this to nanoseconds here (`nanoseconds(Point->time_since_epoch()).count())`), or explicitly set the SteadyTimePoint resolution to nanoseconds (`time_point<steady_clock, nanoseconds>`) -- or both.

Also, size_t is a somewhat unexpected type for a time. That's usually reserved for container indexes.

oontvoo replied:
done :
 - updated SteadyTimePoint to have nanoseconds specified.
 - explicitly call nanoseconds()
 - changed return type to unsigned long long

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,338 @@
+  serializer.writeString("lldb_path", lldb_path);
+  serializer.writeString("cwd", cwd);
+  serializer.writeSizeT("start", stats.start.time_since_epoch().count());
+  serializer.writeSizeT("end", ToNanosecOrZero(stats.end));
```
Why not just omit the field if its not present?

oontvoo replied:
done

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,338 @@
+  // If this entry was emitted at the end of the command-execution,
+  // then calculate the runtime too.
+  if (stats.end.has_value()) {
+    serializer.writeSizeT("command_runtime",
+                          (stats.end.value() - stats.start).count());
```
Why bother computing this (increasing the size of the message and everything) if this can be trivially performed later in the pipeline (if needed)?

oontvoo replied:
done

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,338 @@
+    std::unique_ptr<llvm::telemetry::Config> config,
+    lldb_private::Debugger *debugger) {
+
+  TelemetryManager *ins = new TelemetryManager(std::move(config), debugger);
+
+  return std::unique_ptr<TelemetryManager>(ins);
```
```suggestion
  return std::unique_ptr<TelemetryManager>(new TelemetryManager(std::move(config), debugger));
```

The basic rule of smart pointers is to never have an unowned pointer flying around

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,338 @@
+  entry->SessionId = m_session_uuid;
+
+  for (auto &destination : m_destinations) {
+    llvm::Error err = destination->receiveEntry(entry);
```
question: would it be better to always send the message to each destination and then return all the errors (via `llvm::joinErrors`) if needed?

oontvoo replied:
done - use joinErrors

labath commented on lldb/test/CMakeLists.txt:
```diff
@@ -3,7 +3,7 @@
 # not present.
 if (NOT DEFINED Python3_EXECUTABLE)
-  message(SEND_ERROR
+  message(FATAL_ERROR
```
bad merge?

oontvoo replied:
done

labath reviewed (commented):
> Pavel's suggestion was to split up the patch to make reviewing easier

It was, but this wasn't the kind of separation I had in mind. For me as least, it's very hard to review features in these "horizontal" slices. It's also unclear how (if ever) will this code be tested.

I think it be much better to slice this up vertically, so that (e.g.) the `CommandTelemetryInfo` was added in the same patch that was actually using it.

oontvoo commented:
> > Pavel's suggestion was to split up the patch to make reviewing easier
> 
> It was, but this wasn't the kind of separation I had in mind. For me as least, it's very hard to review features in these "horizontal" slices. It's also unclear how (if ever) will this code be tested.
> 
> I think it be much better to slice this up vertically, so that (e.g.) the `CommandTelemetryInfo` was added in the same patch that was actually using it.

I can split the usage changes to different patches (because they'd be changing differnt files), but spliting the library into different patches by "verticals" seems unnecessarily clustering to me (esp. they're touching the same file, which makes it annoying to resolve conflicts)

oontvoo commented:
> > Pavel's suggestion was to split up the patch to make reviewing easier
> 
> It was, but this wasn't the kind of separation I had in mind. For me as least, it's very hard to review features in these "horizontal" slices. It's also unclear how (if ever) will this code be tested.
> 
> I think it be much better to slice this up vertically, so that (e.g.) the `CommandTelemetryInfo` was added in the same patch that was actually using it.

P.S: I've updated this patch to remove all of the additional base classes of TelemtryInfo, keeping only the LldbBaseTelemtryInfo. Similarly, removed most of the `log...*()` methods from this patch.

PTAL.

oontvoo commented on lldb/include/lldb/lldb-enumerations.h:
```diff
@@ -257,8 +257,8 @@ enum StopReason {
-enum ReturnStatus {
-  eReturnStatusInvalid,
+enum ReturnStatus : int {
+  eReturnStatusInvalid = 0,
```
carry-over comment from previous patch("Why is this needed?")
 This is needed to serialize the ReturnStatus object

oontvoo replied:
when i tried to do `json::Object({"ret_stat", returnStatus})`, I got some build error (ambiguous types or sth like that).

oontvoo replied:
done

oontvoo replied:
done - removed the `:int` and `=0`

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,95 @@
+using ::llvm::telemetry::Destination;
+using ::llvm::telemetry::TelemetryInfo;
+
+static unsigned long long ToNanosec(const SteadyTimePoint Point) {
```
`unsigned long long` is also an unusual type for the number of nanoseconds. The reason for implementing ~all integral types in the generic implementation was so that you can use ~any type here, and things will just work. I'd use `uint64_t` here, but you can also go with `std::chrono::nanoseconds::rep` if you want to be fancy.

oontvoo replied:
done

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,95 @@
+
+  llvm::Error defferedErrs = llvm::Error::success();
+  for (auto &destination : m_destinations) {
+    if (auto err = destination->receiveEntry(entry))
+      deferredErrs = llvm::joinErrors(std::move(deferredErrs), std::move(err));
```
```suggestion
    deferredErrs = llvm::joinErrors(std::move(deferredErrs), destination->receiveEntry(entry));
```

oontvoo replied:
done

labath commented on lldb/test/CMakeLists.txt:
```diff
@@ -108,6 +108,9 @@ endfunction(add_lldb_test_dependency)
 add_lldb_test_dependency(lldb)
 add_lldb_test_dependency(lldb-test)
 
+# Enable Telemetry for testing.
+target_compile_definitions(lldb PRIVATE -DTEST_TELEMETRY)
+
```
This doesn't do what you think it does. It builds the *lldb* binary with the macro definition *if* `LLDB_INCLUDE_TESTS` is defined (because that's the condition that guards cmake [entering this directory](https://github.com/llvm/llvm-project/blob/54665f5252695922dd000f311f82af717f1df0c6/lldb/CMakeLists.txt#L154)).

(Or, to put it differently `target_compile_definitions` always defines a macro for the target it gets as an argument, regardless of where the macro is called from)

labath reviewed (commented):
@JDevlieghere, it looks like the llvm part of this is going to land imminently, so maybe this is a good time to take another look at this. In particular, I expect you'll want to say something about the use of the string "lldb" in identifier names.

labath commented:
> > > Pavel's suggestion was to split up the patch to make reviewing easier
> > 
> > 
> > It was, but this wasn't the kind of separation I had in mind. For me as least, it's very hard to review features in these "horizontal" slices. It's also unclear how (if ever) will this code be tested.
> > I think it be much better to slice this up vertically, so that (e.g.) the `CommandTelemetryInfo` was added in the same patch that was actually using it.
> 
> P.S: I've updated this patch to remove all of the additional base classes of TelemtryInfo, keeping only the LldbBaseTelemtryInfo. Similarly, removed most of the `log...*()` methods from this patch.
> 

Thanks. I think this is better because there's no shortage of things to discuss with the general approach as well. It's true that the vertical slices can cause more merge conflicts, but they should be simple to resolve, and a very big plus is that you can do the reviews (the true bottleneck most of the time) in parallel.

oontvoo commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,95 @@
+
```
done

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,101 @@
+
+private:
+  std::unique_ptr<llvm::telemetry::Config> m_config;
+  Debugger *m_debugger;
```
Sorry, this wasn't supposed to be here. :) I've removed it from the Manager class.
The idea is to have each of the sub-classes of TelemetryInfo carry a Debugger pointer, if applicable.

oontvoo replied:
(done)

oontvoo committed:
Implement LLDB Telemetry (Part 1)

This contains only the concrete implementation of the framework to be used but no usages yet.

This is a subset of PR/98528.

I plan to send a few follow-up patches:
part2 : includes changes in the plugin-manager to set up the plugin stuff (ie., how to create a default vs vendor impl)
part3 (all of the following can be done in parallel):
* part 3_a: define DebuggerTelemetryInfo and related methods to collect data about debugger startup/exit
* part 3_b: define TargetTelemetryInfo and related methods to collect data about debug target(s)
* part 3_c: define CommandTelemetryInfo and related methods to collect data about debug-commands
* part 3_d: define ClientTelemtryInfo and related methods to collect data about lldb-dap/any other client

oontvoo commented:
(accidentally closed the PR - reopening now)

oontvoo commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#include "lldb/Core/Telemetry.h"
```
done

oontvoo committed:
fix order

oontvoo committed:
Merge branch 'llvm:main' into lldb_temetry_s1

oontvoo committed:
Merge branch 'llvm:main' into lldb_temetry_s1

oontvoo committed:
Sync to head and update LLDB's telemetry to match with new llvm::Telemetry changes

oontvoo commented:
@labath Hi, I've updated the patch to sync up with the changes from llvm::Telemetry. (Aslo addressed all the open-questions.) Please have another look when you can. Thanks!

labath commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,92 @@
+
+TelemetryManager::TelemetryManager(
+    std::unique_ptr<llvm::telemetry::Config> config,
+    lldb_private::Debugger *debugger)
+    : m_config(std::move(config)), m_debugger(debugger),
```
It looks like you expect the debugger pointer to always be valid. Could it be a reference? (it means the TelemetryManager object will not be copyable/assignable, but maybe we don't even want that?)

labath replied:
SG

labath reviewed (commented):
I think I don't see all of the updates you've made and the PR has the "Processing updates" forever-spinner. The same thing happened to @cmtice's PR, so you may want to check with her how she got the PR back on track.

oontvoo committed:
put Telemetry sources behind cmake condition-variable

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,100 @@
+
+namespace lldb_private {
+
+using llvm::telemetry::Destination;
```
Ok, I'll just use the full name  then - maybe that's better than creating aliases ...

oontvoo replied:
So putting everthng under `lldb_private::telemetry`?

oontvoo replied:
done.

oontvoo commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,92 @@
+    LLDB_LOG(GetLog(LLDBLog::Object),
+             "Failed to generate random bytes for UUID: {0}", ec.message());
+    // fallback to using timestamp + debugger ID.
+    ret = std::to_string(
+              std::chrono::steady_clock::now().time_since_epoch().count()) +
+          "_" + std::to_string(debugger->GetID());
```
done

oontvoo committed:
Addressed review comments:
- use full names rather than alias
- use formatv

oontvoo commented:
> I think I don't see all of the updates you've made and the PR has the "Processing updates" forever-spinner. The same thing happened to @cmtice's PR, so you may want to check with her how she got the PR back on track.

Seems to be back to normal now! :)

labath reviewed (commented):
I think this looks okay so far. @JDevlieghere, do you have anything to add?

The interesting part will be wiring this up with initialization and all the call sites (which is now complicated by the fact that this code is conditionally compiled.

Before you get too much into the plugin manager, let me say that I *don't* think it makes sense emulating the "normal" plugin setup for this functionality. It's true that this is meant to be pluggable in a way, so calling it *a* plugin sounds reasonable. However, it's a very different kind of a plugin from all of our other plugins. The complexity of those plugins comes from the fact that we have *multiple* implementations (classes) of those plugins, which can create an *arbitrary number* of instances (objects) of that class, at *arbitrary points in time* (e.g. each time we open an object file).

Pretty much none of this applies here. As I understand it, we're going to have *at most one implementation* of the plugin (at most one *active* instance at least), this implementation is going to have *at most one instance* (since we're sharing it between debugger objects), and we're going to be creating it directly *at startup*.

With that in mind, I think it'd be sufficient to have something like `TelemetryManager::SetInstance()`, and have that be called from the `Initialize()` static method of the plugin. (None of the fancy CreateInstance methods, callbacks, etc.)

oontvoo committed:
Update lldb/source/Core/Telemetry.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

oontvoo committed:
Update lldb/source/Core/Telemetry.cpp

Co-authored-by: Pavel Labath <pavel@labath.sk>

oontvoo commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+struct LldbBaseTelemetryInfo : public llvm::telemetry::TelemetryInfo {
+  EventStats stats;
+
+  std::optional<ExitDescription> exit_desc;
```
We're going to have 5 types of entries.I think the exit status  (or more accurately "return status" for the command-info case) is applicable to the first 4 entries:
 - debugger-info  (whether LLDB terminates successfully)
 - target-info  (whether the main executable exits successfully)
 - command-info (whether the command executed successfully)
 - client-request-info (whether the request completed successfully)
 - misc-info (optionall)


So that's 4 out of 5, which is why I thought it made sense to hoist it into the common class.

JDevlieghere replied:
Okay, this answers my previous question. It seems like the meaning of the exit code is sufficiently different (debugger and target have a process exit code, which is meaningful based on the OS) but commands and client request infos share nothing with that. It seems like separate enums would be appropriate for the latter. In other words, I think `debugger-info` and `target-info` each can have an exit code field (or an ExitDescription) member. The other 3 should have their own dedicated enum.

oontvoo replied:
ok -ll remove the field from `LLDBBaseTelemetry`. (But will leave the `ExitDescription` struct decl since the two of following PRs will use it) 
SG?

labath replied:
I think it'd be better to remove the struct completely for now. It'll be easier to figure out the exact meaning/description for it in the context of the PR which uses it.

oontvoo replied:
done - removed the ExistDescription

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+//===-- Telemetry.h ----------------------------------------------*- C++
+//-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
```suggestion
//===-- Telemetry.h -------------------------------------------------------===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
```

(I believe we had an RFC about not doing the `*- C++ -*` anymore)

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+
+namespace lldb_private {
+
+struct LldbEntryKind : public ::llvm::telemetry::EntryKind {
+  static const llvm::telemetry::KindType BaseInfo = 0b11000;
+};
```
```suggestion
struct LLDBEntryKind : public ::llvm::telemetry::EntryKind {
  static const llvm::telemetry::KindType BaseInfo = 0b11000;
};
```
```suggestion
struct LldbEntryKind : public ::llvm::telemetry::EntryKind {
  static const llvm::telemetry::KindType BaseInfo = 0b11000;
};
```

I don't think we use `Lldb` anywhere.

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+struct EventStats {
+  // REQUIRED: Start time of an event
+  SteadyTimePoint start;
+  // OPTIONAL: End time of an event - may be empty if not meaningful.
+  std::optional<SteadyTimePoint> end;
```
The `std::optional` already conveys that it's optional. 
```suggestion
  /// End time of an event - may be empty if not meaningful.
  std::optional<SteadyTimePoint> end;
```

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+
+/// Various time (and possibly memory) statistics of an event.
+struct EventStats {
+  // REQUIRED: Start time of an event
+  SteadyTimePoint start;
```
```suggestion
  /// Start time of an event
  SteadyTimePoint start;
```
This says `REQUIRED` but there's a default constructor that doesn't initialize it?

oontvoo replied:
done

labath replied:
I think what caught Jonas's eye (though he probably doesn't know it) is that this class was very protobuf-y. Let me try to unpack that. We don't have the (understandable if you know the background, but strange if you don't) situation of fields that are *optional* at the transport layer but *required* at the application layer elsewhere. And I don't think we need that here either because the transport layer is part of the downstream/vendor implementation, so we can just use regular c++ conventions for the application layer (if a type is std::optional, it's optional; if it's not, it's not).

That's one part of the issue. The second part is question whether the "required-ness" of a field can be enforced syntactically (through a constructor which intializes it). That's usually a nice property though it doesn't work in all situations, and I don't think it's that useful for objects which are plain data carriers (no internal logic). It also only works if *every* constructor initializes it, so the fact that you've added a constructor which does that, but then also kept the default one which does not (well, it initializes it, but to zero), doesn't really help there.

I'm saying this because this overload set is basically the same as what you'd get if you specified no explicit constructors -- you could still do the same thing, just with curly braces (`EventStats{}` default initializes everything, `EventStats{start}` initializes the first member, `EventStats{start, end}` initializes both). So, if this is the behavior you want, you can just delete all of them. If you want to enforce the requiredness (which might be nice, but I don't think it's necessary -- Jonas may disagree), then you should  delete the default constructor -- but then you also need to use these classes differently.

oontvoo replied:
Sorry for the confusion. Should ahve added more comment on this.
The "REQUIRED" here doesn't mean it has to be set at construction time .It just means it should be set at some point before the TelemetryInfo struct is sent off.

In any case, I've removed the `EventStats` and moved the start/end time fields directly into the TelemetryInfo struct.

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+                                                std::chrono::nanoseconds>;
+
+/// Various time (and possibly memory) statistics of an event.
+struct EventStats {
```
[begin bikeshedding] 
Given the current implementation, I don't think there are "stats".  How about just `EventData` or `Data`. IIUC, this class is meant to represent data that all the LLDB TelemetryInfo will contain. Is there a reason that even needs to be its own class? 
[end bikeshedding]

oontvoo replied:
done (moved the fields into the TelemetryInfo class)

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+using SteadyTimePoint = std::chrono::time_point<std::chrono::steady_clock,
+                                                std::chrono::nanoseconds>;
+
+/// Various time (and possibly memory) statistics of an event.
```
```suggestion
/// Various time statistics of an event.
/// TODO: Extend with memory statistics. 
```
I know what you mean, but the current comment is just confusion: does it contain memory or not? I think a `TODO` is how such things are usually expressed. Or you could omit it and put the todo where you have the `TBD`.

oontvoo replied:
done

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,93 @@
+      : start(start), end(end) {}
+};
+
+/// Describes the exit signal of an event.
```
I don't know what `the exit signal of an event` means. Do you mean the exit signal of a process? Of LLDB itself?

oontvoo replied:
it's used by a few subclasses of the TelemetryInfo. (Comment below)

oontvoo replied:
To answer the orignal question, this is used to describe both the exit of LLDB and a process (or main executable).
I've removed the "signal" from the comment.

JDevlieghere reviewed (commented):
I left a bunch of nits but the core looks good.

oontvoo commented on lldb/source/Core/CMakeLists.txt:
```diff
@@ -17,6 +17,15 @@ if (LLDB_ENABLE_CURSES)
 endif()
 
 # TODO: Add property `NO_PLUGIN_DEPENDENCIES` to lldbCore
+set(LLDB_CORE_SRCS
+
+)
+
```
done

oontvoo committed:
remove unused cmake var

oontvoo committed:
Update lldb/include/lldb/Core/Telemetry.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

oontvoo committed:
Update lldb/include/lldb/Core/Telemetry.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

oontvoo committed:
Update lldb/include/lldb/Core/Telemetry.h

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

oontvoo committed:
addressed review comments

oontvoo committed:
removed 'Required' from comment

oontvoo committed:
remove explicit =0 on enum

oontvoo committed:
move code under lldb_private::telemetry

oontvoo commented:
@JDevlieghere  Hi, do you have any further comment/feedback  on this? If not, I plan to merge the PR this afternoon. Thanks!

oontvoo committed:
Merge branch 'main' into lldb_temetry_s1

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,74 @@
+
+  Debugger *debugger;
+
+  // For dyn_cast, isa, etc operations.
```
Everywhere else we use `/// LLVM RTTI support.`

JDevlieghere commented on lldb/include/lldb/Core/Telemetry.h:
```diff
@@ -0,0 +1,74 @@
+  void serialize(llvm::telemetry::Serializer &serializer) const override;
+};
+
+/// The base Telemetry manager instance in LLDB
```
Nit: missing period.

oontvoo replied:
done

JDevlieghere commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,69 @@
+    serializer.write("end_time", ToNanosec(end_time.value()));
+}
+
+static std::string MakeUUID(lldb_private::Debugger *debugger) {
```
`s/lldb_private:://`, goes for the whole file.

oontvoo replied:
done

oontvoo replied:
(all addressed in https://github.com/llvm/llvm-project/pull/126757)

JDevlieghere commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,69 @@
+  if (auto ec = llvm::getRandomBytes(random_bytes, 16)) {
+    LLDB_LOG(GetLog(LLDBLog::Object),
+             "Failed to generate random bytes for UUID: {0}", ec.message());
+    // fallback to using timestamp + debugger ID.
```
`s/fallback/Fallback/`

oontvoo replied:
done

JDevlieghere commented on lldb/source/Core/Telemetry.cpp:
```diff
@@ -0,0 +1,69 @@
+}
+
+TelemetryManager::TelemetryManager(
+    std::unique_ptr<llvm::telemetry::Config> config)
```
You're using `using ::llvm::telemetry::Foo` for some other things but not for the Config. Why not use `using namespace llvm::telemetry` instead?

oontvoo replied:
done

kazutakahirata commented:
@oontvoo I've checked in 622b8edfc2485b21802778d1f4ae7aed3bbbb40e to fix a warning.  I've put `[[maybe_unused]]` instead of removing the function, guessing that you might use this function in subsequent patches.

jimingham commented:
This PR caused the llvm Standalone bot build to fail, for instance:

https://ci.swift.org/view/all/job/llvm.org/view/LLDB/job/lldb-cmake-standalone/1328/consoleText

Can someone fix that build problem?

mgorny commented:
`PARTIAL_SOURCES_INTENDED` doesn't work there since it isn't passed to `llvm_add_library()`, and `llvm_process_sources()` is called again there.

That said, I would personally lean towards putting the whole file around an `#if` rather than disabling the QA check. In the end, `PARTIAL_SOURCES_INTENDED` is mostly used when there are two targets built in a single CMake file, rather than for conditional sources.

mgorny commented:
Filed #126710 to fix passing `LLVM_BUILD_TELEMETRY` in standalone builds, and #126715 to fix building with telemetry disabled.

oontvoo commented:
Related follow-up : [PR/126746](https://github.com/llvm/llvm-project/pull/126746)  (for excluding telemetry code)
